lib/control-connection.js

```diff
@@ -417,23 +418,21 @@ class ControlConnection extends events.EventEmitter {
 
   async _refreshControlConnection(hostIterator) {
 
-    if (this.options.sni) {
-      this.connection = this._borrowAConnection(hostIterator);
-    }
-    else {
-      try { this.connection = this._borrowAConnection(hostIterator); }
-      catch(err) {
-
-        /* NODEJS-632: refresh nodes before getting hosts for reconnect since some hostnames may have
-         * shifted during the flight. */
-        this.log("info", "ControlConnection could not reconnect using existing connections.  Refreshing contact points and retrying");
-        this._contactPoints.clear();
-        this._resolvedContactPoints.clear();
-        await Promise.all(this.options.contactPoints.map(name => this._parseContactPoint(name)));
-        const refreshedContactPoints = Array.from(this._contactPoints).join(',');
-        this.log('info', `Refreshed contact points: ${refreshedContactPoints}`);
-        await this._initializeConnection();
+    try { this.connection = this._borrowAConnection(hostIterator); }
+    catch(err) {
+
+      this.log("info", "ControlConnection could not reconnect using existing connections", err);
+      if (this.options.cloud) {
+
+        this.log("info", "Refreshing host IDs for cloud nodes from metadata service");
+        await cloud.init(this.options);
+        this.log("info", `Current host IDs: ${this.options.contactPoints}`);
       }
+
+      this.log("info", "Re-initializing the control connection");
+      this._contactPoints.clear();
+      this._resolvedContactPoints.clear();
+      await this.init(true);
```

Comment:
We still get different behaviour in the SNI case (which was the original goal of the if/else logic above)... it's just that now that logic lives in this.init() rather than inline.